src/RevCPU.cc

```diff
@@ -52,12 +48,12 @@ RevCPU::RevCPU( SST::ComponentId_t id, const SST::Params& params )
 
   // Derive the simulation parameters
   // We must always derive the number of cores before initializing the options
-  numCores = params.find<unsigned>( "numCores", "1" );
-  numHarts = params.find<uint16_t>( "numHarts", "1" );
+  numCores = params.find<uint32_t>( "numCores", "1" );
```

Comment:
these should probably stay as uint16s - no need to have more than 65K cores per CPU or more than 65K harts per core

Reply:
> these should probably stay as uint16s - no need to have more than 65K cores per CPU or more than 65K harts per core

Then we cannot produce an error when the user requests `>= 65536`, because the test will always be false. A `uint16_t` can never be `>= 65536` so it will end up being some value `mod 65536` which we will accept and which the user didn't ask for. By using `uint32_t` we can at least catch and enforce when the user exceeds `65535`. The original code compared a `uint16_t >= 65536` which is always `false`. We allow `uint32_t` input when getting the parameter, but then we enforce that it isn't too large:
```c++
  // Make sure someone isn't trying to have more than 65536 harts per core
  if( numHarts > _MAX_HARTS_ ) {
    output.fatal( CALL_INFO, -1, "Error: number of harts must be <= %" PRIu32 "\n", uint32_t{ _MAX_HARTS_ } );
  }
```
Actually right now, the `_MAX_HARTS_` is defined as `4096`:
```c++
common/include/RevCommon.h:#define _MAX_HARTS_    4096
```
We should parse as `uint32_t` or `uint64_t`, then enforce the limit. If we parse as `uint16_t` we will always be under `65536` and won't catch anything larger as an error.

Reply:
Also, there is no space / performance advantage to expressing `numHarts` and `numCores` as `uint16_t`; it might even be slower than using the more "native" word sizes of `uint32_t` and `uint64_t`. We should use the more native word sizes `uint32_t` or `uint64_t`, and enforce the limits as we already do above, and then keep them at a native word size. If we use `uint16_t` to read the value, we will get a mod 65536 value which the user didn't want and we won't know what they asked for. Even if SST takes the `params.find` template type into account and raises overflow errors if the value is too large, that's less helpful of an error message than our limit check above.

Reply:
Also, if you literally mean "Make sure someone isn't trying to have **more than** 65536 harts per core", then to allow `65536` requires `uint32_t` because `uint16_t` cannot represent `65536`.

The declaration type of a variable does not imply intention about its permitted range. I can have a `uint64_t` variable and still enforce that its value is `<= 65536`. But if I have a `uint16_t` variable, I cannot tell whether the user asked for `> 65536` and raise an error for it.

The type and the allowable range of a parameter are separate but partially related concerns.